pkg/apply/event/event.go

```diff
@@ -75,92 +81,73 @@ type ErrorEvent struct {
 	Err error
 }
 
-//go:generate stringer -type=ApplyEventType
-type ApplyEventType int
+//go:generate stringer -type=ActionGroupEventType
+type ActionGroupEventType int
 
 const (
-	ApplyEventResourceUpdate ApplyEventType = iota
-	ApplyEventCompleted
+	Started ActionGroupEventType = iota
+	Finished
 )
 
+type ActionGroupEvent struct {
+	GroupName string
+	Action    ResourceAction
+	Type      ActionGroupEventType
+}
+
 //go:generate stringer -type=ApplyEventOperation
 type ApplyEventOperation int
 
 const (
-	ServersideApplied ApplyEventOperation = iota
+	NoneApply ApplyEventOperation = iota
+	ServersideApplied
 	Created
 	Unchanged
 	Configured
-	Failed
 )
 
 type ApplyEvent struct {
-	Type       ApplyEventType
-	Operation  ApplyEventOperation
-	Object     *unstructured.Unstructured
 	Identifier object.ObjMetadata
+	Operation  ApplyEventOperation
+	Resource   *unstructured.Unstructured
 	Error      error
 }
 
-//go:generate stringer -type=StatusEventType
-type StatusEventType int
-
-const (
-	StatusEventResourceUpdate StatusEventType = iota
-	StatusEventCompleted
-)
-
 type StatusEvent struct {
-	Type     StatusEventType
-	Resource *pollevent.ResourceStatus
+	Identifier       object.ObjMetadata
+	PollResourceInfo *pollevent.ResourceStatus
+	Resource         *unstructured.Unstructured
+	Error            error
 }
 
-//go:generate stringer -type=PruneEventType
-type PruneEventType int
-
-const (
-	PruneEventResourceUpdate PruneEventType = iota
-	PruneEventCompleted
-	PruneEventFailed
-)
-
 //go:generate stringer -type=PruneEventOperation
 type PruneEventOperation int
 
 const (
-	Pruned PruneEventOperation = iota
+	NonePrune PruneEventOperation = iota
+	Pruned
 	PruneSkipped
 )
 
 type PruneEvent struct {
-	Type       PruneEventType
+	Identifier object.ObjMetadata
 	Operation  PruneEventOperation
 	Object     *unstructured.Unstructured
-	Identifier object.ObjMetadata
 	Error      error
 }
 
-//go:generate stringer -type=DeleteEventType
-type DeleteEventType int
-
-const (
-	DeleteEventResourceUpdate DeleteEventType = iota
-	DeleteEventCompleted
-	DeleteEventFailed
-)
-
```

Comment:
Just to make sure my understanding is correct. The `DeleteEventType` is removed. Similar for ApplyEventType and PruneEventType. That means for consumers, they can't rely on the DeleteEventFailed to check if failure happens. They need to check the Error field inside each event.

Reply:
Yeah, the idea here is that we no longer have any events that pertains to all resources in the task. So every event of ApplyEventType/PruneEventType/DeleteEventType will belong to a single resource, i.e. the `Identifier` field should always have a value. This also means that there is only one way to check for failures, and that is to check the `Error` field.